auction-server/src/simulator/main.rs

```diff
@@ -0,0 +1,191 @@
+use {
+    litesvm::{
+        types::{
+            FailedTransactionMetadata,
+            SimulatedTransactionInfo,
+        },
+        LiteSVM,
+    },
+    solana_client::{
+        rpc_config::RpcSendTransactionConfig,
+        rpc_response::{
+            Response,
+            RpcResult,
+        },
+    },
+    solana_rpc_client::nonblocking::rpc_client::RpcClient,
+    solana_sdk::{
+        account::{
+            Account,
+            ReadableAccount,
+        },
+        account_utils::StateMut,
+        bpf_loader_upgradeable::UpgradeableLoaderState,
+        commitment_config::CommitmentConfig,
+        pubkey::Pubkey,
+        signature::Signature,
+        transaction::VersionedTransaction,
+    },
+    std::{
+        collections::HashMap,
+        time::Instant,
+    },
+    tokio::sync::RwLock,
+};
+
+pub struct Simulator {
+    sender:      RpcClient,
+    receiver:    RpcClient,
+    pending_txs: RwLock<Vec<(VersionedTransaction, Instant)>>,
+}
+
+struct AccountsConfig {
+    accounts: HashMap<Pubkey, Account>,
+    programs: HashMap<Pubkey, Account>,
+}
+
+impl AccountsConfig {
+    fn new() -> Self {
+        Self {
+            accounts: Default::default(),
+            programs: Default::default(),
+        }
+    }
+
+    fn apply(&self, svm: &mut LiteSVM) {
+        for (key, account) in self.accounts.iter() {
+            svm.set_account(*key, account.clone()).unwrap();
+        }
+        for (key, account) in self.programs.iter() {
+            svm.add_program(
```

Comment:
how does this work with program upgrades? why not just set account data for the program and its executable as well?

Reply:
tried that first, litesvm didn't support it very well. But we can create a PR for that in litesvm package